### PR TITLE
Fixed issue where the scene wasn't centered in the models

### PR DESCRIPTION
**Pull request by JCash**
## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java [hunk 1]
```diff
@@ -670,48 +670,56 @@ private static void findModelNodes(Node node, List<Node> modelNodes) {
...
+        Aabb aabb = new Aabb();
         for (Node root : scene.rootNodes) {
-            ArrayList<Node> modelNodes = new ArrayList<>();
-            findModelNodes(root, modelNodes);
-
-            for (Node modelNode : modelNodes) {
-                center.x += modelNode.local.translation.x;
-                center.y += modelNode.local.translation.y;
-                center.z += modelNode.local.translation.z;
-                count++;
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java:684]: This is not how to calculate a center point, using a flattened tree and local points?! (not sure what I was thinking)

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java [hunk 2]
```diff
@@ -670,48 +670,56 @@ private static void findModelNodes(Node node, List<Node> modelNodes) {
...
-            ArrayList<Node> modelNodes = new ArrayList<>();
-            findModelNodes(root, modelNodes);
-
-            for (Node modelNode : modelNodes) {
-                modelNode.local.translation.x -= center.x;
-                modelNode.local.translation.y -= center.y;
-                modelNode.local.translation.z -= center.z;
-                modelNode.world.translation.x -= center.x;
-                modelNode.world.translation.y -= center.y;
-                modelNode.world.translation.z -= center.z;
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java:704]: And, why alter the local positions?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java [hunk 3]
```diff
@@ -914,6 +922,16 @@ public static void main(String[] args) throws IOException {
         }
         System.out.printf("--------------------------------------------\n");
 
+        System.out.printf("Root Nodes:\n");
+
+        for (Node node : scene.rootNodes) {
+            System.out.printf("  Scene Node: %s  index: %d  id: %d  parent: %s\n", node.name, node.index, ModelImporter.AddressOf(node), node.parent != null ? node.parent.name : "");
+            System.out.printf("      local: id: %d\n", ModelImporter.AddressOf(node.local));
+            ModelImporter.DebugPrintTransform(node.local, 3);
+        }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java:931]: More debug output when debugging a model.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java [hunk 4]
```diff
@@ -670,48 +670,56 @@ private static void findModelNodes(Node node, List<Node> modelNodes) {
         }
     }
 
+    private static void calcCenterNode(Node node, Aabb aabb) {
+        if (node.model != null) {
+            // As a default, we only count nodes with models, as the user
+            // cannot currently see/use the lights or cameras etc that are present in the scene.
+            aabb.expand(node.world.translation.x, node.world.translation.y, node.world.translation.z);
+        }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java:678]: The center realignment is for the users benefit, to make it easier to directly place a model in the scene, where the .model is placed.

So, we only let the models contribute, and not the lights/cameras etc.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java [hunk 5]
```diff
@@ -670,48 +670,56 @@ private static void findModelNodes(Node node, List<Node> modelNodes) {
...
 
+    private static void shiftModelsNode(Node node, ModelImporter.Vec4 center) {
+        node.world.translation.x -= center.x;
+        node.world.translation.y -= center.y;
+        node.world.translation.z -= center.z;
+
+        for (Node child : node.children) {
+            shiftModelsNode(child, center);
+        }
+    }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ModelUtil.java:706]: Once the center of the aabb is calcualted, we can shift the scene there.
In this case we move _all_ the nodes, as we want their relative positions intact.

**Comment by Mulderua** [comment]
It is the same for our problem (

image

**Comment by JCash** [comment]
Hmm, ok. Would you mind sharing that model with me?

**Comment by Mulderua** [comment]
> Hmm, ok. Would you mind sharing that model with me?

I am preparing the project and details that I'll send to you in Discord.

**Review by Jhonnyg — APPROVED** [review]
LGTM!

**Pull request merged**